Fixes #6710 - unicode characters in url parameters

tstrachota opened the pull request:
Original methods to_param defined on resources called name.parameterize
to get rid of url-unsafe characters. This function also strips off unicode
characters.
Another problem was in one unified search function for all parameterization approaches. It used to fail when a name was parameterized to blank string or when a name contained only numbers.

Changes:
- parameterization extracted into a separate module Parameterizable
- each resource defines methods 'to_param' and 'from_param' to ensure correct search
- views updated to always use to_param when building links
- parameterization changed to id+name for resources where it wasn't safe to rely only on names
- default parameter is id

Commit:
Fixes #6710 - unicode characters in url parameters

Original methods to_param defined on resources called name.parameterize
to get rid of url-unsafe characters. This function unfortunately also
stripped off unicode characters.

Changes:
- parameterization extracted into a separate module Parameterizable
- each resource defines methods 'to_param' and 'from_param' to ensure correct search
- views updated to always use to_param when building links
- parameterization changed to id+name for resources where it wasn't safe to rely only on names
- default parameter is id

tstrachota commented on app/controllers/application_controller.rb:
```diff
@@ -121,18 +121,26 @@ def model_of_controller
   # example:
-  # @host = Host.find_by_name params[:id]
+  # before_filter :find_by_name, :only => %w{edit update destroy}
   def find_by_name
```
I kept the original method name for backwards compatibility with plugins but it doesn't make much sense now. Should I rename it, what do you think?

dLobatog replied:
Probably now you should use `friendly_id` after this commit was merged. https://github.com/theforeman/foreman/commit/8b737c9c7648b3726dadb3b2e4708fcb43af02a8

ohadlevy commented on app/models/concerns/parameterizable.rb:
```diff
@@ -0,0 +1,50 @@
+    included do
+      def to_param
+        # remove characters unsafe for urls, keep unicode ones
+        "#{id}-#{name}".gsub(/[!*'();:@&=+$,\/?%#\[\]]/, '').chomp('-')
```
did you consider using http://apidock.com/rails/ActiveSupport/Inflector/transliterate ?

tstrachota replied:
I did but `transliterate` replaces uknown characters with `?` which could result in having plenty of question marks instead of names (e.g. in Chinese).
From the tests I did, Rails encodes unicode characters in urls well with %. Browser translates the encoded characters back in the address bar so there's no loss of readability.

ohadlevy commented on app/models/fact_name.rb:
```diff
@@ -1,4 +1,6 @@
 class FactName < ActiveRecord::Base
+  include Parameterizable::ByIdName
```
byname or by idname ?

tstrachota replied:
I intentionally changed it to id-name because I didn't find any format validation for `FactName` and considered it not safe for usage in urls.
Is the name format guaranteed by any other conditions (e.g. puppet restrictions)?

ohadlevy commented on app/models/lookup_key.rb:
```diff
@@ -1,5 +1,6 @@
 class LookupKey < ActiveRecord::Base
   include Authorizable
+  include Parameterizable::ById
```
i wonder if thats not going to be the default regardless ? e.g. do we need to explicitly define that in every model?

ohadlevy commented on config/initializers/active_record_extensions.rb:
```diff
@@ -2,4 +2,5 @@ class ActiveRecord::Base
   include StripWhitespace
-end
\ No newline at end of file
+  include Parameterizable::ById
```
if this is included, I assume we dont need it in other models (such as lookupkey) ?

tstrachota replied:
Correct. That line in lookupkeys was superfluous.

GregSutcliffe commented:
Needs rebasing

mbacovsky commented:
Rebased

ohadlevy commented:
needs another rebase it seems :)

mbacovsky commented:
Rebased

dLobatog commented on app/controllers/application_controller.rb:
```diff
@@ -121,18 +121,26 @@ def model_of_controller
-
-  # searches for an object based on its name and assign it to an instance variable
-  # required for models which implement the to_param method
+  # Searches for a record based on it's parameterized version. If the model
```
`it's` -> `its`

dLobatog commented:
This needs rebasing now that we use `friendly_id`, I expect some changes will be necessary.

dLobatog commented:
Note from discussion on IRC - NestedAncestryCommon is overriding to_param for taxonomies and host groups at least, so that a root host group with all unicode chars will become ``#{id}-`. I guess we should use the unicode chars. This might be happening at other places since `Parametrize::ById` is loaded very early compared to concerns or class methods.

tstrachota commented:
Updated. I fixed some more occurrences of `to_param` that remained without attention after the rebases. All should be fine now.
Thanks for catching that @eLobato.

dLobatog commented:
@tstrachota Awesome, could you just add one or two test to check unicode is 'translated' in the URL? For models with NestedAncestryCommon and models without it

tstrachota commented:
@eLobato Updated. I added tests for `parameterize` and `to_param` on hostgroups.

dLobatog commented:
@tstrachota There's a bug introduced, host groups and other stuff that uses ancestry, is now encoding the URLs without dashes to separate parent/children, like: 

`http://localhost:3000/hostgroups/1673-adcmonfrontierfrontdev/edit`

It should be 

`http://localhost:3000/hostgroups/1663-adcmon-frontier/edit`

It does it for non unicode stuff and unicode stuff, for example

`http://localhost:3000/hostgroups/1767-acron%D1%A0%D1%A1%D1%A2%D1%A3%D1%A4%D1%A5%D1%A6%D1%A7%D1%A8%D1%A9%D1%AA%D1%AB%D1%AC%D1%AD%D1%AE%D1%B7%D1%B8%D1%B9%D1%BA%D1%BB%D1%BC%D1%BD%D1%BE%D1%BF/edit`

Please fix and add a tests for these dashes and I think we're good to go :rocket:  :+1:

tstrachota commented:
@eLobato Parameterization updated, changes reflected in tests:
https://github.com/theforeman/foreman/pull/1685/files#diff-a620d136f1e20e88384db07470f9c124R12
https://github.com/theforeman/foreman/pull/1685/files#diff-ec15c79bfabb56e6ef525d4f7c65808bR334

dLobatog commented:
@tstrachota Thanks, it's looking fine. Merging in 3..2..1 :bomb:

dLobatog commented:
Merged as https://github.com/theforeman/foreman/commit/e768c9761528c64fcc289c391091fcce63e82ff8, thanks @tstrachota
